Item processing error for work_type

> ITEM@1957-3125
 [2017-05-26T11:59:36]: [ERROR] Failed processing: undefined local variable or method `work_type' for #<Item:0x007ffac8bdaef0> :
	 /usr/local/deploy-rails/dspace2hydra/lib/bag/item.rb:112:in `build_metadata_hash'
	/usr/local/deploy-rails/dspace2hydra/lib/bag/item.rb:13:in `metadata'
	/usr/local/deploy-rails/dspace2hydra/lib/work/migration_strategy/base.rb:36:in `process_bag_metadata'
	/usr/local/deploy-rails/dspace2hydra/lib/work/migration_strategy/single_work.rb:13:in `process_bag'
	dspace2hydra.rb:129:in `block in <main>'
	dspace2hydra.rb:114:in `each'
	dspace2hydra.rb:114:in `<main>'
ITEM@1957-3125 [2017-05-26T11:59:36]: undefined local variable or method `work_type' for #<Item:0x007ffac8bdaef0> : /usr/local/deploy-rails/dspace2hydra/lib/bag/item.rb:112:in `build_metadata_hash'
	/usr/local/deploy-rails/dspace2hydra/lib/bag/item.rb:13:in `metadata'
	/usr/local/deploy-rails/dspace2hydra/lib/work/migration_strategy/base.rb:36:in `process_bag_metadata'
	/usr/local/deploy-rails/dspace2hydra/lib/work/migration_strategy/single_work.rb:13:in `process_bag'
	dspace2hydra.rb:129:in `block in <main>'
	dspace2hydra.rb:114:in `each'
	dspace2hydra.rb:114:in `<main>'

--- a/config/default.yml
+++ b/config/default.yml
@@ -81,13 +81,6 @@ custom_nodes:
       type: Array
     method: Mapping::Community.lookup_community_names
     value_from_node_property: collection_handles
-  keyword:
-    field:
-      name: keyword
-      property: "%{work_type}.%{field_name}"
-      type: Array
-    method: Mapping::Keyword.unprocessed
-    value: data-migration
   license:
     field:
       name: license

--- a/config/etd.yml
+++ b/config/etd.yml
@@ -81,13 +81,6 @@ custom_nodes:
       type: Array
     method: Mapping::Community.lookup_community_names
     value_from_node_property: collection_handles
-  keyword:
-    field:
-      name: keyword
-      property: "%{work_type}.%{field_name}"
-      type: Array
-    method: Mapping::Keyword.unprocessed
-    value: data-migration
   license:
     field:
       name: license

--- a/lib/bag/item.rb
+++ b/lib/bag/item.rb
@@ -109,7 +109,7 @@ class Item
     # so creating a new document from the mutated temp_xml is effective, albeit hackish
     remaining_xml = Nokogiri::XML.parse temp_xml.to_xml
     # TODO: consider making this error configurable by Item type
-    raise StandardError, "#{work_type} : #{metadata_xml_path} unhandled nodes:\n#{remaining_xml.to_xml}" unless remaining_xml.root.children.empty?
+    raise StandardError, "#{work_type_config.dig('work_type')} : #{metadata_xml_path} unhandled nodes:\n#{remaining_xml.to_xml}" unless remaining_xml.root.children.empty?
     h
   end
 

--- a/lib/metadata/qualifier.rb
+++ b/lib/metadata/qualifier.rb
@@ -14,13 +14,24 @@ module Metadata
       @work_type = work_type_config['work_type']
       @node_config = node_config
       @node_config['field'] ||= {}
-      @config = node_config['qualifiers'][qualifier]
-      @config['field'] ||= {}
+      set_config(work_type_config, node_config, qualifier)
       @value = value
     end
 
     def default?
       @qualifier == 'default'
+    end
+
+    def set_config(work_type_config, node_config, qualifier)
+      @config = node_config.dig('qualifiers', qualifier)
+      raise_config_error(work_type_config, node_config, qualifier) if @config.nil?
+      @config['field'] ||= {}
+    end
+
+    def raise_config_error(work_type_config, node_config, qualifier)
+      message = "#{work_type_config.dig('work_type')} :  Missing #{node_config.dig('field', 'name')}.#{qualifier} qualifier configuration."
+      @logger.fatal(message)
+      raise message
     end
 
     def value_add_to_migration

--- a/spec/fixtures/mocks/.config.yml
+++ b/spec/fixtures/mocks/.config.yml
@@ -21,6 +21,7 @@ hydra_endpoint:
     url: "/users/sign_in"
     username: "admin_user"
     password: "admin_user_password"
+    authentication_token: "blahblah"
 
     # If the url is redirected to another authentication system, such as CAS, the configurations set below should be
     # relative to the form that is presented to the user for typing in their username and password.

--- a/spec/fixtures/mocks/default.yml
+++ b/spec/fixtures/mocks/default.yml
@@ -24,6 +24,9 @@ hydra_endpoint:
     # which form on the page will be used for manipulation and submission.. as well as grabbing the CSRF token from it.
     form_action: "/concern/default_works?locale=en"
 
+  new_child_work:
+    form_action: "/concern/parent/%{parent_id}/default_works?local=en"
+
   workflow_actions:
     auto_advance_work: true
     url: "/concern/workflow_actions/%{id}?local=en"

--- a/spec/lib/hydra_endpoint/configuration_spec.rb
+++ b/spec/lib/hydra_endpoint/configuration_spec.rb
@@ -37,6 +37,10 @@ RSpec.describe HydraEndpoint::Configuration do
     expect(subject.admin_sets_url).to eq URI.join(server_domain, config.dig('admin_sets', 'url'))
   end
 
+  it 'has a authentication_header' do
+    expect(subject.authentication_header).to eq 'D2H-AUTHENTICATION' => 'admin_user|blahblah'
+  end
+
   it 'has a login_url' do
     expect(subject.login_url).to eq URI.join(server_domain, config.dig('login', 'url'))
   end

--- a/spec/lib/hydra_endpoint/hydra_endpoint_spec.rb
+++ b/spec/lib/hydra_endpoint/hydra_endpoint_spec.rb
@@ -64,6 +64,15 @@ RSpec.describe HydraEndpoint::Server do
       end
     end
 
+    context 'when submitting a new child work' do
+      let(:work_response) { subject.submit_new_child_work(bag, data, 'parent_id') }
+      it 'can submit_new_child_work' do
+        expect(subject).to receive(:cache_data).with(data, bag.item_cache_path, 'data')
+        expect(subject).to receive(:publish_work).with(data, subject.new_child_work_action('parent_id'), {}).and_return(mock_hydra_endpoint_response)
+        expect(work_response).to eq mock_hydra_endpoint_response
+      end
+    end
+
     context 'when advancing a work through a workflow' do
       let(:work_response) { subject.advance_workflow(mock_hydra_endpoint_response, headers) }
       it 'can advance_workflow' do

--- a/spec/lib/loggable_spec.rb
+++ b/spec/lib/loggable_spec.rb
@@ -9,6 +9,11 @@ RSpec.describe Loggable do
     end
   end
   subject { klass.new }
+  it '#log_to_summary' do
+    subject.log_to_summary('message')
+    logger = Logging.logger['summary_log']
+    expect(logger.appenders.count).to eq 1
+  end
   it '#start_logging_to' do
     logger = subject.start_logging_to('/tmp/test', {})
     expect(logger.appenders.count).to eq 1

--- a/spec/lib/metadata/custom_node_spec.rb
+++ b/spec/lib/metadata/custom_node_spec.rb
@@ -67,6 +67,10 @@ RSpec.describe Metadata::CustomNode do
     it 'can return collection_handles' do
       expect(custom_node.collection_handles).to match_array(['1957/43909', '1957/4'])
     end
+
+    it 'can return metadata' do
+      expect(custom_node.metadata).to be_a_kind_of Hash
+    end
   end
   context 'when get_configuration fails' do
     subject { custom_node }

--- a/spec/lib/metadata/qualifier_spec.rb
+++ b/spec/lib/metadata/qualifier_spec.rb
@@ -74,6 +74,13 @@ RSpec.describe Metadata::Qualifier do
         expect(subject.run_method).to eq 'executed test_method with value grimey and arg1,arg2'
       end
     end
+
+    context 'with a missing configuration' do
+      let(:qualifier) { 'doesntexist' }
+      it 'logs a fatal message and raises an exception' do
+        expect { subject.run_method }.to raise_error
+      end
+    end
   end
 
   context 'when get_configuration fails' do

